### ISSUE TITLE
v-ref does not seem to work certain times

### ISSUE THREAD

**Issue by jatins** [yyx990803/vue#714]
I have this parent component:
```
<div id="app" >
		<div v-component="search" v-ref="env"></div>
               <div v-on="click: takeAction">
			<a>okay</a>
		</div>
</div>
```

and the corresponding VueModel instance:
```js
var app = new Vue({
	el: '#app',
	data: {
	},
	methods: {
		takeAction: function() {
			var env = this.$.env,
			var text = env.text,
			console.log(text);		
		}
	},
	components: {
		search: require('./search')
	},
});
```

The `search` component has an input field with `v-model="text"`. I want to access the text from child component in parent. The problem is that it just seems to work sometimes, while other times, `this.$.env` is just `undefined`. 

Also, is there some other way to get only some data from child environment (without `inherit:true` which will leak all data from parent to child)? I though of using events but let's say I have multiple search components on the page, and I want the `text` field from all of them, events wouldn't work in such case. 

Maybe something like `exports: {}` on the child component which exposes an api to other components as well as parent.

**Comment by TMiguelT** [comment]
I can't help you with the first problem because you'll need to give a more specific situation where `this.$.*` is undefined. 

As for the second issue, yes there is a way to get data from the child. What you do is pass in a parent variable to the child using `v-with` (or use `paramAttributes`), which then modifies that variable. In your case it would use that variable in a `v-model`. The Vue guide talks about data passing with v-with [here](http://vuejs.org/guide/components.html#Passing_Down_Individual_Properties).

And [here's a JSFiddle](http://jsfiddle.net/L7t4x85L/) that passes a parent field to a child which then manipulates it. The textbox at the top is a child component, but when you change it, the text underneath also changes, which is in the parent scope, meaning that it's all working as you want.

```html
<my-input v-with="bound: varA"></my-input>
<p>{{varA}}</p>
```

```javascript
new Vue({
    el: "body",
	data:{
		varA: "",
	},
	components:{
		"my-input": {
			template: '<input type="text" v-model="bound">',
		}
	}
});
```

**Comment by jatins** [comment]
Oh, I am using `0.11.4` which and bidirectional bindings when using `v-with` was introduced in `0.11.5`

While I guess I can use v-with in this case, I think it'd be a really nice if we could expose some sort of an api for each component.

**Comment by TMiguelT** [comment]
Well all data is already "exposed". You can just access `this.$.child.<someProperty>` as you did in your example. It's just that that way is annoying and imperative in comparison to `v-with`. How would you access child data with your idea? And you should be able to update to 0.11.5 without any breaking changes.

**Comment by jatins** [comment]
Yeah, that's what I am saying. With `this.$.child` , _all child data_ is exposed. It gives me _read, write access to all child data_ which can be potentially dangerous.
However, I am suggesting a commonjs style functionality for components.

So, let's say I have a data variable `foo` in child component.
Then with `this.$.child`, I can read `foo`, as well as _modify_ `foo`. But if I could expose a function `getFoo()` from child component, that way I could the parent could read `foo` from child component, but the parent can not accidentally modify it.

Not only that, it gives me a fine grained control on what I really want to expose to other components.

**Comment by cheapsteak** [comment]
I've run into something similar.
The ref was inside a nested component.
Using `v-with`, `ComponentAAA` passed data to `ComponentBBB`, and some of that data was passed from `ComponentBBB` to `ComponentCCC`. 
When `ComponentAAA`'s data changed, `this.$.ComponentCCC` inside `ComponentBBB` became `undefined`.  
As a workaround, I'm manually saving a reference to the reference in the `compiled` hook in `ComponentBBB`:   
```
compiled: function () {
    this.ComponentCCC = this.$.ComponentCCC;
  },
```

Probably not the best way to go but things need to get done :\

**Comment by yyx990803** [comment]
There was a fix for a related `v-ref` bug in 0.11.5, not sure if it has addressed your issue though.

**Comment by yyx990803** [comment]
Closing since there's no actionable repro to investigate - feel free to reopen if this pops up again.
